### PR TITLE
lightning: remove the unnecessary `offset` field in the key encoding for duplicate detection

### PR DESCRIPTION
**Pull request by gozssky**
### What problem does this PR solve?

Part of https://github.com/pingcap/tidb/issues/29588.
Split from https://github.com/pingcap/tidb/pull/30203.

Problem Summary:

The `rowID` that allocated by lightning is guaranteed to be unique, so we can remove the unnecessary `offset` field in the key encoding for duplicate detection.

### What is changed and how it works?

Remove the unnecessary `offset` field in the key encoding for duplicate detection.

I tested about 5 million rows for TPCC's  order_line schema.
```sql
CREATE TABLE `order_line` (
  `ol_o_id` int(11) NOT NULL,
  `ol_d_id` int(11) NOT NULL,
  `ol_w_id` int(11) NOT NULL,
  `ol_number` int(11) NOT NULL,
  `ol_i_id` int(11) NOT NULL,
  `ol_supply_w_id` int(11) DEFAULT NULL,
  `ol_delivery_d` datetime DEFAULT NULL,
  `ol_quantity` int(11) DEFAULT NULL,
  `ol_amount` decimal(6,2) DEFAULT NULL,
  `ol_dist_info` char(24) DEFAULT NULL,
  PRIMARY KEY (`ol_w_id`,`ol_d_id`,`ol_o_id`,`ol_number`) /*T![clustered_index] NONCLUSTERED */
) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
```
|                                        | index-engine | data-engine |
| -------------------------------------- | ------------ | ----------- |
| `duplicate-resolution="none"`           | 31 M          | 194 M        |
| `duplicate-resolution="record"`(master)  | 71 M          | 238 M        |
| `duplicate-resolution="record"`(this PR) | 52 M          | 218 M        |

The results show that the write amplification of the index engine and data engine is significantly reduced.

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: br/pkg/lightning/backend/local/engine.go [hunk 1]
```diff
@@ -1008,25 +1006,29 @@ func (w *Writer) appendRowsSorted(kvs []common.KvPair) error {
...
-		w.batchSize += int64(len(encodedKvs[i].Key) + len(encodedKvs[i].Val))
+		keySize := keyAdapter.EncodedLen(kvs[i].Key)
+		w.batchSize += int64(keySize + len(kvs[i].Val))
+		totalKeySize += keySize
+	}
+	w.batchCount += len(kvs)
+	// noopKeyAdapter doesn't really change the key,
+	// skipping the encoding to avoid unnecessary alloc and copy.
+	if _, ok := keyAdapter.(noopKeyAdapter); !ok {
+		buf := make([]byte, 0, totalKeySize)
```
Comment (glorv) [inline br/pkg/lightning/backend/local/engine.go:1022]: I think we can put this but in `Writer` for reusing to avoid alloc for each call.

**Review by glorv — COMMENTED** [review]
rest LGTM

File: br/pkg/lightning/backend/local/engine.go (hunk 1 above)
Comment (gozssky) [reply br/pkg/lightning/backend/local/engine.go]: Done.

**Review by glorv — APPROVED** [review]
LGTM

**Comment by gozssky** [comment]
@kennytm Could you please take a look? Thanks

**Review by Little-Wallace — APPROVED** [review]
LGTM

**Pull request merged**